### DISCUSSION TITLE
16 Tone Generator Dual MiniDexed work in progress

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#504] [Show and tell]
I'm messing around with combining two MiniDexed units to give me 16 Tone Generators.  Of course, this can be done already using MIDI and treating them as two independent MIDI-controlled devices, but I want a single UI to control all 16 Tone Generators.

Here is a short rather uninspiring, demonstration showing that in principle it works: https://www.youtube.com/watch?v=SjmsP43-J_8

In this video, the main MiniDexed changes the voice on TG1 from brass to strings, then switches to (remote) TG9 and changes the voice there too.  Both TG1 and TG9 are listening on MIDI channel 1 so will play in response to the MIDI events from the keyboard.

There are a number of constraints:
* Both MiniDexed devices must have the same set of voice ROMs and performances.
* It relies on a one-way serial MIDI link from the Main to the expander.
* If the UI on the expander unit is used, then the two devices will get out of sync with each other.
* The audio outputs remain independent, so it isn't possible to mix the two paths internally - an external mixer is required, although in principle I guess you could have one device on each of the stereo channels...
* Certain effects remain "global" even if they are different in performances on the two boards.
* Remote voice editing is not possible.
* The Main (real) TGs are 1-8 and the remote TGs are 9-16, but on the remote unit they are still shown as 1-8 if a screen is present.  So main TG9 maps onto remote TG1, etc.

The idea is to allow two Pis to run controlled from one UI.

The configuration required to do this:
Main MiniDexed:
MIDIThru=ttyS1,ttyS1
MIDIAutoVoiceDumpOnPC=0
TGExpanders=8
TGExpanderStart=1

Expander MiniDexed:
(MIDIThru disabled)
MIDIAutoVoiceDumpOnPC=0
TGExpanders=0
TGExpanderStart=9

Anyone fancy designing a 16-voice "Performance"?  It would have to be installed in two parts - one for each TG1-8 on each board, and all voices would need to respond to the same MIDI channel.

Thoughts? Is this something of interest? Is there any real point considering most of what I'm doing could be done via MIDI?

Current branch of code for this is here: https://github.com/diyelectromusic/MiniDexed/tree/TGExpander

Kevin

**Comment by Banana71** [comment]
I could put together an orchestra for you tomorrow on 16 TG's in 2 performances.  

Couldn't a Raspberry Pi 4 handle 16 TG's at once?

**Reply by diyelectromusic** [reply]
> Couldn't a Raspberry Pi 4 handle 16 TG's at once?

No idea tbh :)  Currently the cores of a Pi4 are used as follows:
Core 0: main IO, etc
Core 1: main sound mixing/etc and 2 TGs
Cores 2 and 3: three TGs each

Kevin

**Reply by Banana71** [reply]
is the division the same for RPi 2 and 3?

**Reply by diyelectromusic** [reply]
I think so - aren't they all quad core above V1?

**Reply by pik33** [reply]
These RPi4 cores are about 2x more powerful than RPi3. Modern RPi2 is in reality underclocked (to 900 MHz) Rpi3.. so if RPi2 can do 8 TGs, RPi4 should do 16.

**Comment by Banana71** [comment]
> Main MiniDexed: MIDIThru=ttyS1,ttyS1 MIDIAutoVoiceDumpOnPC=0 TGExpanders=8 TGExpanderStart=1
> 
> Expander MiniDexed: (MIDIThru disabled) MIDIAutoVoiceDumpOnPC=0 TGExpanders=0 TGExpanderStart=9

I don't understand the Parameter TGExpanders.:
Why ist on main the value "8" and on the second Pi "0"?
I expected something like master and slave(s).

**Reply by diyelectromusic** [reply]
> I don't understand the Parameter TGExpanders.: Why ist on main the value "8" and on the second Pi "0"?

It is the number of remote expanders for that device to manage.  The main one is managing 8 remote TGs, the expander device has no further expanders.

I wasn't quite sure how best to articulate the difference, and this is just the first attempt.  It might end up completely different if I carry it on :)

Kevin

**Reply by Banana71** [reply]
Master and Slave(n)

**Reply by diyelectromusic** [reply]
> Master and Slave(n)

No, I don't use those terms in professional or technical contexts anymore.  Controller and peripheral, main and branch, or in this case main and remote expander is fine.

I mean I wasn't sure how to articular how many expanders and so on - I'm still working out how I want it to work and how flexible I want it to be.  It might not be a simple one main and one expander for example... it might be nice to be able to hook up 8 Pi V1s...

Kevin

**Comment by Banana71** [comment]
> Anyone fancy designing a 16-voice "Performance"? It would have to be installed in two parts - one for each TG1-8 on each board, and all voices would need to respond to the same MIDI channel.

@diyelectromusic 16 parts at the same time very quickly becomes mush of sound. Can you describe in more detail what you want?

[16 Voice Performance.zip](https://github.com/probonopd/MiniDexed/files/11641828/16.Voice.Performance.zip)

Peter

**Reply by diyelectromusic** [reply]
I don't have anything in mind tbh, I was just idly tinkering and wondered if anyone would be interested and what they'd do with it if so! :)

But thanks - I'll take a look.

Kevin

**Reply by pik33** [reply]
16 voices may be good for keyboard and/or dynamic splitting. I saw (or maybe rather heared..) a performance on a synth (maybe Yamaha Montage? it was several years ago) that played one violin at low velocity, and then going to a violin ensemble at higher velocities.

**Comment by diyelectromusic** [comment]
> Couldn't a Raspberry Pi 4 handle 16 TG's at once?

I now have a build for a RPi4 and 16 TGs here: https://github.com/diyelectromusic/MiniDexed/tree/Pi4-16TG

It might need some fiddling around with mixing levels when going at "full blast", but I've not really looked in detail yet.  Its just an experiment at the moment.

Kevin

**Reply by akivaatwood** [reply]
Is there a pre-compiled build? I can't seem to find it if it's there

Thanks

**Reply by diyelectromusic** [reply]
No sorry, just the branch of code at the moment - I was really just seeing what might chime with people as interesting :)

Kevin

**Reply by akivaatwood** [reply]
Thanks -- I'll have to look into building it myself (when I have some time)

**Comment by diyelectromusic** [comment]
I've now written up all my experiments so far with 16 tone generators here: https://diyelectromusic.wordpress.com/2023/07/21/16-tone-generator-minidexed/

This includes a PCB I had built to link to Pis together :)

Kevin

![IMG_7278](https://github.com/probonopd/MiniDexed/assets/68612569/2252985b-ab6a-4f39-8790-47a698611225)

**Comment by diyelectromusic** [comment]
Ok, I have a new WIP that combines the PI4 16 TGs with the Pi 2/3 8+8 TGs: https://github.com/diyelectromusic/MiniDexed/tree/16TGExpander

This is perhaps ready for some proper testing now.  Note the config settings have changed to something that I hope is little more sensible:
TGLocal=1, 8 or 16 - really only relevant on a Pi 4 to select between 8 and 16.
TGRemote=0..8 - for a Pi 1,2,3 configures the menu to support an expander.
TGLocalStart=1 or 9 - defines which TG (if used as an expander) this board starts with.

So for normal operation - i.e. if none of the above are specified at all, then
TGLocal=1 for PI1 or 8 for PI 2,3,4
TGRemote=0
TGLocalStart=1

To configure a PI2/3 setup for dual operation:
First Pi:
TGLocal=8
TGRemote=8
TGLocalStart=1

Second Pi:
TGLocal=8
TGRemote=0
TGLocalStart=9

To configure a Pi 4 for 16 TG operation:
TGLocal=16

It should be possible to configure a couple of PI1s for two TGs - one on each Pi - too, but I've not really looked at that in detail yet.  Ideally I'd like to be able to pass UI messages around so that 8 Pi 1s could give us back the 8 TGs... but that is perhaps for a future experiment.

There are still some limitations.  In theory TGLocal could be used to trim down the number of TGs for a PI 2/3 but I've not allowed that yet as it ought to have some more sensible spreading of TGs over cores...

Also the UI has expanders starting from TG 9 even on a Pi 1 at the moment, as the menu structures are all hard-coded.

Oh and you still can't edit voices for expander TGs and I don't plan to add that either.

Anyway, this is where I'm at with it.  Let me know if you think this is interesting.  I'll keep tinkering with it regardless.

Kevin

**Reply by diyelectromusic** [reply]
Ok, so I've been giving this a bit of a trashing and I think it largely seems ok to me.  I've tested with the following configurations:
Pi 4 - default 8 TGs
Pi 4 - 16 TGs
Pi 3 - default 8 TGs
Dual Pi 3 - 8+8 TGs
Pi 1 - single TG

I've also tested it with dual Pi 1s but that isn't working atm so I need to have a bit of a think.  In the back of my mind I'd quite like to allow several Pi 1s to act together to give a multi-TG setup, but I'm not sure if it's really worth the effort to manage the UI properly or not yet.  I'll have a think.

But I think 16 TGs as an option either for a single Pi 4 or for a dual set of Pi 3s seems to work fine as far as I can tell.

But I could really do with someone else giving it a go before submitting a PR.  That is assuming anyone is interested in having it as an option of course!  Is anyone interested in this as a feature!? :)

Kevin

**Reply by Banana71** [reply]
Hi Kevin,
I am very intrested in the PI 4 variant.
My PI 4 would be very happy about a little more workload :-)

**Reply by diyelectromusic** [reply]
Well it should build fine for the Pi 4 and then just set TGLocal=16 in your minidexed.ini.  Oh and then you'll need a fair bit of cut-and-paste in your performance.ini if you want anything interesting going on, but then I probably don't need to tell you about performance.ini :)

Do let me know if you manage to give it a go!

Kevin

**Reply by diyelectromusic** [reply]
Dual Pi 1 should be working too now.  In fact, I think I should be able to stack up to 8 Pi 1s together, but haven't got as far as testing that yet :)

Kevin

**Comment by diyelectromusic** [comment]
Have you run out of Pi 3s or 4s?  Don't worry.  If you have a spare 8 Raspberry Pi V1 boards lying around then you could try this.

(Yes I know it's pointless, but you know, sometimes you just have to try things just to see if it works :)) 

Kevin

https://www.youtube.com/watch?v=rghPRLvLk8s

**Reply by probonopd** [reply]
Impressive setup. Must have been worth a fortune duing the supply crisis.

I wonder whether the Raspberry Pi 5 can do this (or more?) out of the box. Well, I guess we'll be able to find out soon-ish ;-)

**Reply by diyelectromusic** [reply]
Well seeing as the Pi 4 can apparently relatively painlessly do 16, I suspect the Pi 5 could probably go for 32 ;)

But we can't do anything until it is supported in Circle, which from what I've read about the changes for the Pi 5 looks like it might actually be quite a task...

Kevin

**Comment by Andy2No** [comment]
I'd like to give this a try, maybe starting with 16TG on a Pi4, since that seems simplest.  I'm not confident in being able to build it from source code though.  Any chance of a ready built version, ready to go onto an SD card?

FWIW, there is an option on the TX81Z to use a pair of them with alternating voice allocation, so one gets the odd numbered notes and one gets the even ones - e.g. playing a chord with 4 notes, each TX81Z module would get two notes to play.  I don't know if it's practical to do it that way but it would make it easier to test that it's working properly - e.g. cable them up so that each Pi feeds one ear.  RCA leads are good for doing that sort of thing.

A good reason for having so many tone generators would be to allow notes with long tails /  release periods to continue playing while new notes are played... I'm not clear how the number of Tone Generators relates to polyphony, in this context.

**Reply by diyelectromusic** [reply]
Well I need to do some work to update it to the latest codebase - then I think if I send it in as a PR then it will start auto building... :)

Kevin

**Comment by RK11111111111** [comment]
VCV rack had a modular FM synth unit.  That would be kind of interesting, but DX7 doesn't really do resynthesis.  Pico ADK looks like a neat module, it is a pi built with a DAC and analog inputs.

In theory you could create a digitally controlled modular synth that allows automatic reloading or rerouting of audio.  Maybe using the ethernet to send digital streams rather than analog cables.  I think would be fun, so that you could reload patches through sysex rather than  plugging cables.

**Reply by probonopd** [reply]
Speaking of resynthesis, I can imagine the day when we attach a microphone, record some sound (like on a sampler), and let some AI magic create a DX7 patch matching it closely. Ah, [wait](https://fcaspe.github.io/ddx7/)...

**Reply by RK11111111111** [reply]
There is even source available to do that.

https://github.com/spiegelib/vst-fm-sound-match

It's useless python developer code though.

**Reply by probonopd** [reply]
Useless? I'd say it's pretty good:
https://spiegelib.github.io/spiegelib/examples/fm_sound_match_pages/fm_sound_match_listen.html#fm-sound-match-listen

If you have a sample of something and want a DX7 patch that you can use on MiniDexed, I'd say it's definitely worth a try.

But let's move this to its [own thread](https://github.com/probonopd/MiniDexed/discussions/615).
